src/qibo/models/error_mitigation.py

```diff
@@ -592,8 +632,340 @@ def apply_randomized_readout_mitigation(
+
+        qubit_map (list, optional): the qubit map. If None, a list of range of circuit's qubits is used. Defaults to ``None``.
+        backend (qibo.backends.abstract.Backend, optional): the backend to be used in the execution.
+            If None, it uses the global backend. Defaults to ``None``.
+
+    Returns:
+        float: The mitigated expectation value of the observable.
+    """
+    if backend is None:  # pragma: no cover
+        backend = GlobalBackend()
+
+    if "ncircuits" in readout:
+        circuit_result, circuit_result_cal = apply_randomized_readout_mitigation(
+            circuit, noise_model, nshots, readout["ncircuits"], backend
+        )
+    else:
+        circuit_result = _circuit_conf(
+            circuit, qubit_map, noise_model, nshots, backend=backend
+        )
+
+    if "response_matrix" in readout:
+        circuit_result = apply_resp_mat_readout_mitigation(
+            circuit_result,
+            readout["response_matrix"],
+            readout.get("ibu_iters", None),
+        )
+
+    exp_val = circuit_result.expectation_from_samples(observable)
+
+    if "ncircuits" in readout:
+        exp_val /= circuit_result_cal.expectation_from_samples(observable)
+
+    return exp_val
+
+
+def sample_clifford_training_circuit(
+    circuit,
+    backend=None,
+):
+    """Samples a training circuit for CDR by susbtituting some of the non-Clifford gates.
+
+    Args:
+        circuit (:class:`qibo.models.Circuit`): circuit to sample from.
+        backend (:class:`qibo.backends.abstract.Backend`, optional): backend to be used
+            in the execution. If ``None``, it uses :class:`qibo.backends.GlobalBackend`.
+            Defaults to ``None``.
+
+    Returns:
+        :class:`qibo.models.Circuit`: The sampled circuit.
+    """
+    from qibo.quantum_info import (  # pylint: disable=import-outside-toplevel
+        random_clifford,
+    )
+
+    if backend is None:  # pragma: no cover
+        backend = GlobalBackend()
+
+    non_clifford_gates = [
+        (i, gate)
+        for i, gate in enumerate(circuit.queue)
+        if not gate.clifford and not isinstance(gate, gates.M)
+    ]
+
+    if not non_clifford_gates:
+        raise_error(ValueError, "No non-Clifford gate found, no circuit sampled.")
+
+    sampled_circuit = circuit.__class__(**circuit.init_kwargs)
+
+    for i, gate in enumerate(circuit.queue):
+        if isinstance(gate, gates.M):
+            for q in gate.qubits:
+                gate_rand = gates.Unitary(
+                    random_clifford(1, backend=backend, return_circuit=False),
+                    q,
+                )
+                gate_rand.clifford = True
+                sampled_circuit.add(gate_rand)
+            sampled_circuit.add(gate)
+        else:
+            if i in [index for index, _ in non_clifford_gates]:
```

Comment:
Better not to build the same list every iteration of the loop
```suggestion
            if i in non_clifford_gates_indices:
```